### PR TITLE
UCP/WIREUP: Fix data race

### LINKED ISSUE #3642: shm_ib/test_ucp_tag_mt.send_recv/1 fails on boo31
```
15:59:46 [----------] 1 test from shm_ib/test_ucp_tag_mt
15:59:46 [ RUN      ] shm_ib/test_ucp_tag_mt.send_recv/1
16:00:12 [boo31:19595:0:19595] ucp_proxy_ep.c:214  Assertion `proxy_ep->uct_ep != ((void *)0)' failed
16:00:16 ==19595== brk segment overflow in thread #1: can't grow to 0x4800000
16:00:16 ==19595== (see section Limitations in user manual)
16:01:21 
16:01:21 /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucs/debug/assert.c: [ ucs_fatal_error_message() ]
16:01:21       ...
16:01:21        33     }
16:01:21        34 
16:01:21        35     ucs_handle_error(message_buf);
16:01:21 ==>    36     abort();
16:01:21        37 }
16:01:21        38 
16:01:21        39 void ucs_fatal_error_format(const char *file, unsigned line,
16:01:21 
16:02:11 ==== backtrace (tid:  19595) ====
16:02:11  0 0x000000000003c018 ucs_fatal_error_message()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucs/debug/assert.c:36
16:02:11  1 0x000000000003c11f ucs_fatal_error_format()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucs/debug/assert.c:52
16:02:11  2 0x000000000001d08d ucp_proxy_ep_replace()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucp/core/ucp_proxy_ep.c:214
16:02:11  3 0x000000000005d7a9 ucp_signaling_ep_am_short()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucp/wireup/signaling_ep.c:56
16:02:11  4 0x0000000000052699 uct_ep_am_short()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/uct/api/uct.h:2100
16:02:11  5 0x0000000000052699 ucp_tag_send_inline()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucp/tag/tag_send.c:159
16:02:11  6 0x0000000000052699 ucp_tag_send_nb_inner()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucp/tag/tag_send.c:193
16:02:11  7 0x0000000000052699 ucp_tag_send_nb()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../src/ucp/tag/tag_send.c:180
16:02:11  8 0x0000000000696211 test_ucp_tag::send_nb()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/ucp/test_ucp_tag.cc:180
16:02:11  9 0x0000000000696fde test_ucp_tag::send_b()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/ucp/test_ucp_tag.cc:213
16:02:11 10 0x000000000066b504 test_ucp_tag_mt_send_recv_Test::test_body()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/ucp/test_ucp_tag_mt.cc:76
16:02:11 11 0x000000000066b9d1 test_ucp_tag_mt_send_recv_Test::test_body()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/ucp/test_ucp_tag_mt.cc:67
16:02:11 12 0x000000000051407a ucs::test_base::run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/test.cc:276
16:02:11 13 0x00000000005140bd ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/test.cc:302
16:02:11 14 0x00000000005072ed HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3562
16:02:11 15 0x00000000004fd5ec testing::Test::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3635
16:02:11 16 0x00000000004fd8b8 testing::TestInfo::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3812
16:02:11 17 0x00000000004fda7e testing::TestCase::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3930
16:02:11 18 0x00000000004fda7e testing::TestCase::total_test_count()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3864
16:02:11 19 0x00000000004fda7e ???()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3929
16:02:11 20 0x00000000004fe063 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:5802
16:02:11 21 0x00000000004fe063 ???()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:5800
16:02:11 22 0x0000000000506e7d HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:3562
16:02:11 23 0x00000000004fcb39 testing::UnitTest::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:5416
16:02:11 24 0x00000000004fcb39 ???()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/gtest-all.cc:5416
16:02:11 25 0x00000000005078c5 main()  /scrap/jenkins/workspace/hpc-ucx-pr/label/boo31/worker/1/contrib/../test/gtest/common/main.cc:101
16:02:11 26 0x000000347981ed1d __libc_start_main()  ???:0
16:02:11 27 0x00000000004e3889 _start()  ???:0
16:02:11 =================================
```

http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/label=boo31,worker=1/10429/console

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Moves sending of `UCP_WIREUP_MSG_ACK` from an async thread (e.g. UD's `uct_ud_iface_timer` that progress data) to `ucp_worker_progress` function.

## Why ?

Fixes #3642
Also, it prevents the calling of UCT AM routines from different threads at the same time

## How ?

Register the function that sends `UCP_WIREUP_MSG_ACK` in UCT worker.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10473/ for details (Mellanox internal link).

File: src/ucp/wireup/wireup.c [hunk 1]
```diff
@@ -473,13 +487,11 @@ ucp_wireup_process_reply(ucp_worker_h worker, const ucp_wireup_msg_t *msg,
...
-        ucs_trace("ep %p: sending wireup ack", ep);
-        memset(rsc_tli, -1, sizeof(rsc_tli));
-        status = ucp_wireup_msg_send(ep, UCP_WIREUP_MSG_ACK, 0, rsc_tli);
-        if (status != UCS_OK) {
-            return;
-        }
+        /* Send `UCP_WIREUP_MSG_ACK` from progress function
+         * to avoid calling UCT routines from an async thread */
+        uct_worker_progress_register_safe(worker->uct,
+                                          ucp_wireup_send_msg_ack, ep,
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: need to clean this while endpoint is destroyed

File: src/ucp/wireup/wireup.c [hunk 2]
```diff
@@ -435,11 +435,25 @@ ucp_wireup_process_request(ucp_worker_h worker, const ucp_wireup_msg_t *msg,
     }
 }
 
+static UCS_F_NOINLINE unsigned ucp_wireup_send_msg_ack(void *arg)
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: why do we need UCS_F_NOINLINE ?

File: src/ucp/wireup/wireup.c [hunk 3]
```diff
@@ -435,11 +435,25 @@ ucp_wireup_process_request(ucp_worker_h worker, const ucp_wireup_msg_t *msg,
...
+    ucp_ep_h ep = (ucp_ep_h)arg;
+    ucp_rsc_index_t rsc_tli[UCP_MAX_LANES];
+    ucs_status_t status;
+
+    /* Send ACK without any address, we've already sent it as part of the request */
+    ucs_trace("ep %p: sending wireup ack", ep);
+
+    memset(rsc_tli, UCP_NULL_RESOURCE, sizeof(rsc_tli));
+    status = ucp_wireup_msg_send(ep, UCP_WIREUP_MSG_ACK, 0, rsc_tli);
+    return (status != UCS_OK) ? 0 : 1;
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: return status==UCS_OK

File: src/ucp/wireup/wireup.c (hunk 1 above)
Comment (dmitrygladkov) [reply src/ucp/wireup/wireup.c]: done

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10487/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_tag_offload.cc [hunk 4]
```diff
@@ -329,9 +329,14 @@ class test_ucp_tag_offload_multi : public test_ucp_tag_offload {
...
     {
         se.connect(&receiver(), get_ep_params());
-        // Need to send twice, since the first message may not enable hashing
-        // (num_active_iface on worker is increased after unexpected offload handler)
-        send_recv(se, tag, 2048);
+        // Need to send twice:
+        // 1. to ensure that wireup's UCT iface has been closed and
+        //    it is not considered for num_active_iface on worker
+        //    (message has to be less than `UCX_TM_THRESH` value)
+        // 2. to activate tag oofload
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_offload.cc]: offload

File: test/gtest/ucp/test_ucp_tag_offload.cc [hunk 5]
```diff
@@ -329,9 +329,14 @@ class test_ucp_tag_offload_multi : public test_ucp_tag_offload {
...
+        //    (num_active_iface on worker is increased after unexpected offload
+        //     hadnler and message has to be greater than `UCX_TM_THRESH` value)
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_offload.cc]: handler

File: test/gtest/ucp/test_ucp_tag_offload.cc [hunk 6]
```diff
@@ -329,9 +329,14 @@ class test_ucp_tag_offload_multi : public test_ucp_tag_offload {
...
         se.connect(&receiver(), get_ep_params());
-        // Need to send twice, since the first message may not enable hashing
-        // (num_active_iface on worker is increased after unexpected offload handler)
-        send_recv(se, tag, 2048);
+        // Need to send twice:
+        // 1. to ensure that wireup's UCT iface has been closed and
+        //    it is not considered for num_active_iface on worker
+        //    (message has to be less than `UCX_TM_THRESH` value)
+        // 2. to activate tag oofload
+        //    (num_active_iface on worker is increased after unexpected offload
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_offload.cc]: num_active_iface is increased when any message is received on any iface. Tag hashing is done when we have more than 1 active ifaces

File: test/gtest/ucp/test_ucp_tag_offload.cc (hunk 4 above)
Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_offload.cc]: fixed

File: test/gtest/ucp/test_ucp_tag_offload.cc (hunk 5 above)
Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_offload.cc]: fixed

File: test/gtest/ucp/test_ucp_tag_offload.cc (hunk 6 above)
Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_offload.cc]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10489/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10494/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10497/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe could you review pls?

File: test/gtest/ucp/test_ucp_tag_offload.cc [hunk 7]
```diff
@@ -329,9 +329,16 @@ class test_ucp_tag_offload_multi : public test_ucp_tag_offload {
...
+        // Need to send twice:
+        // 1. to ensure that wireup's UCT iface has been closed and
+        //    it is not considered for num_active_iface on worker
+        //    (message has to be less than `UCX_TM_THRESH` value)
+        // 2. to activate tag ofload
+        //    (num_active_ifaces on worker is increased when any message
+        //     is received on any iface. Tag hashing is done when we have
+        //     more than 1 active ifaces and message has to be greater
+        //     than `UCX_TM_THRESH` value)
+        send_recv(se, tag, 8);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_offload.cc]: why did the change in wireup expose this issue?

Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_offload.cc]: @yosefe it seems that this is a timing issue:
1. before this fix - it destroys UD iface in the async thread when reply is received
2. now - it destroys UD iface in the application thread that calls `ucp_worker_progress`. So, if we send a message that > `UCX_TM_THRESH`, it has two active ifaces (UD iface for wireup + another iface(e.g. RC) and tag offload engine decides to use TAG hashing)

**Pull request merged**